kotlin/README.md

```diff
@@ -0,0 +1,616 @@
+from adjacent case blocks with a blank line:
+
+```
+private fun parsePropertyValue(propName: String, token: Token) {
+	  when (token) {
+		    is Token.ValueToken ->
+			  callback.visitValue(propName, token.value)
+		
+		    Token.LBRACE -> { // …
+        }
+    }
+}
+```
+
+Put short branches on the same line as the condition, without braces.
+
+```
+when (foo) {
+    true -> bar() // good 
+    false -> { baz() } // bad 
+}
+```
+
+### Method call formatting:
+In long argument lists, put a line break after the opening parenthesis. Indent 
+arguments by 4 spaces. Group multiple closely related arguments on the same line.
+
+```
+drawSquare( 
+    x = 10, y = 10, 
+    width = 100, height = 100, 
+    fill = true )
+```
+
+Put spaces around the = sign separating the argument name and value.
+
+### Lambda formatting:
+In lambda expressions, spaces should be used around the curly braces, as well as 
+around the arrow which separates the parameters from the body. 
+If a call takes a single lambda, it should be passed outside of parentheses whenever
+possible.
+
+```
+list.filter { it > 10 }
+```
+
+If assigning a label for a lambda, do not put a space between the label and the opening
+ curly brace:
+
+```
+fun foo() { 
+    ints.forEach list@{ 
+        // ... 
+    } 
+}
+```
+
+When declaring parameter names in a multiline lambda, put the names on the first line, 
+followed by the arrow and the newline:
+
+```
+appendCommaSeparated(properties) { 
+    prop -> val propertyValue = prop.get(obj)  // ... 
+}
+```
+
+If the parameter list is too long to Dt on a line, put the arrow on a separate line:
+
+```
+foo { 
+    context: Context, 
+    environment: Env 
+    -> 
+    context.configureEnv(environment) 
+}
+```
+
+### Documentation comments:
+For longer documentation comments, place the opening /** on a separate line and begin 
+each subsequent line with an asterisk:
+
+```
+/** 
+* This is a documentation comment 
+* on multiple lines. 
+*/
+```
+
+Short comments can be placed on a single line:
+
+```
+/** This is a short documentation comment. */
+```
+
+String templates:
+Don't use curly braces when inserting a simple variable into a string template. 
+Use curly braces only for longer expressions. 
+
+```
+println("$name has ${children.size} children")
+```
+
+### Type aliases:
+If you have a functional type or a type with type parameters which is used multiple times 
+in a codebase, prefer deDning a type alias for it: 
+
+```
+typealias PersonIndex = Map<String, Person>
+```
+
+### Using conditional statements:
+Prefer using the expression form of try, if and when. Examples:
+
+```
+return if (x) foo() else bar()
+
+return when(x) { 
+    0 -> "zero" 
+    else -> "nonzero" 
+}
+```
+
+The above is preferable to:
+
+```
+if (x) 
+    return foo()
+else 
+    return bar() 
+
+when(x) { 
+    0 -> return "zero" 
+    else -> return "nonzero" 
+}
+```
+
+#### if versus when
```

Comment:
There shouldn't be any performance cost using `when` instead of `if/else` as the compiler reuses `if/else`, `switch` and ternary operator in these cases and sometimes could be more easily readable. Idk what you guys think about this one

Reply:
I agree with the proposed document. I'd use `if` for binary conditions that won't change over time. If there's a binary condition but it's likely that more cases will appear in the future, then I'd use `when`.